File: Java/src/main/java/com/gildedrose/item/ItemFactory.java
package com.gildedrose.item;

public class ItemFactory {

    public CustomisedItem customiseItem(Item item) {
        if (item.name.equals("Sulfuras, Hand of Ragnaros")) {
            return new Sulfuras(item);
        } else if (item.name.equals("Aged Brie")) {
            return new AgedBrie(item);
        } else if (item.name.equals("Backstage passes to a TAFKAL80ETC concert")) {
            return new BackstagePassesItem(item);
        } else {
            return new StandardItem(item);
        }
    }
}

File: ruby/lib/gilded_rose/item_factory.rb
module GildedRose
  class ItemFactory
    SPECIAL_ITEM_KLASSES = {
      "Aged Brie" => AgedBrieItem,
      "Backstage passes to a TAFKAL80ETC concert" => BackstagePassesItem,
      "Sulfuras, Hand of Ragnaros" => SulfurasItem,
    }

    GENERIC_ITEM_KLASS = GenericItem

    def self.create_item(name:, sell_in:, quality:)
      klass = SPECIAL_ITEM_KLASSES.fetch(name, GENERIC_ITEM_KLASS)
      klass.new(name, sell_in, quality)
    end
  end
end

